File: c++/src/Stack.cpp
#include "Stack.h"

namespace wtf {

  Stack::Stack()  {

  }

  size_t Stack::size() const {
    return mList.size();
  }

  bool Stack::isEmpty() const {
    return mList.isEmpty();
  }

  void Stack::push(const int& val) {
    mList.pushFront(val);
  }

  int Stack::pop() {
    return mList.popFront();
  }

  int Stack::peek() const {
    return mList.head();
  }
}

File: elixir/epi_book/lib/chapter_8/stack.ex
defmodule Chapter8.Stack do
  defstruct head: []

  @moduledoc """
  Stack implementation.
  """

  @doc """
  Construct an empty stack.
  """
  def new(), do: %Chapter8.MaxStack{}

  @doc """
  Add an element to the stack.
  """
  def push(stack, val) do
    %Chapter8.MaxStack{head: [val|stack.head]}
  end

  @doc """
  Remove an element from the stack.
  """
  def pop(stack) do
    if stack.head == [] do
      {%Chapter8.MaxStack{head: []}, nil}
    else
      [head|tail] = stack.head
      {%Chapter8.MaxStack{head: tail}, head}
    end
  end

  @doc """
  Return the first element in the stack
  """
  def peek(stack) do
    List.first(stack.head)
  end

end
